sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,895 @@
+    ///
+    /// Returns the index if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    #[inline(always)]
+    pub fn position(&self, target_slot: Slot) -> Option<usize> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot.cmp(&target_slot).reverse())
+            .ok()
+    }
+
+    /// Returns a `&[SlotHashEntry]` view into the underlying data.
+    ///
+    /// The constructor (either the safe path that called `parse_and_validate_data` or
+    /// the unsafe `new_unchecked`) is responsible for ensuring the slice is big enough
+    /// and properly aligned.
+    #[inline(always)]
+    fn as_entries_slice(&self) -> &[SlotHashEntry] {
+        if self.len == 0 {
+            return &[];
+        }
+
+        debug_assert!(self.data.deref().len() >= NUM_ENTRIES_SIZE + self.len * ENTRY_SIZE);
+
+        let entries_ptr =
+            unsafe { self.data.deref().as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry };
+        unsafe { core::slice::from_raw_parts(entries_ptr, self.len) }
+    }
+}
+
+// Implementation block specific to the safe Ref version
+impl<'a> SlotHashes<'a> {
+    /// Creates a `SlotHashes` instance by safely borrowing data from an `AccountInfo`.
+    ///
+    /// This function verifies that:
+    /// - The account key matches the `SLOTHASHES_ID`
+    /// - The data contains a valid length prefix
+    /// - The data is sufficiently large to hold the indicated number of entries
+    #[inline(always)]
+    pub fn from_account_info(account_info: &'a AccountInfo) -> Result<Self, ProgramError> {
+        if account_info.key() != &SLOTHASHES_ID {
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        let data_ref = account_info.try_borrow_data()?;
+        // Ensure the byte slice is suitably aligned for `SlotHashEntry`
+        if (data_ref.as_ptr() as usize) % mem::align_of::<SlotHashEntry>() != 0 {
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        let num_entries = Self::parse_and_validate_data(&data_ref)?;
+
+        Ok(unsafe { Self::new_unchecked_ref(data_ref, num_entries) })
+    }
+}
+
+/// Reads the entry count directly from the beginning of a byte slice **without validation**.
+/// (This is identical to the struct method get_entry_count_unchecked,
+/// added here for discoverability with other unchecked fns)
+///
+/// # Safety
+///
+/// This function is unsafe because it performs no checks on the input slice.
+/// The caller **must** ensure that:
+/// 1. `data` contains at least `NUM_ENTRIES_SIZE` (8) bytes.
+/// 2. The first 8 bytes represent a valid `u64` in little-endian format.
+/// 3. Calling this function without ensuring the above may lead to panics
+///    (out-of-bounds access) or incorrect results.
+#[inline(always)]
+pub unsafe fn get_entry_count_from_slice_unchecked(data: &[u8]) -> usize {
```

Comment:
Perhaps we won't need this methods outside the struct if the struct is generic.

Reply:
After unifying impls in earlier commits, removed these methods in abdc30371194e6a164c1798fe96436dd02decd9e